<h1>Multi Model Server Container for Differential Deep Learning</h1>

This notebook demonstrates how to build and use a custom Docker container for serving with Amazon SageMaker that leverages on the <strong>Multi Model Server (MMS)</strong> and <strong>sagemaker-inference-toolkit</strong> libraries for serving models through Amazon SageMaker's endpoints.
We will also see how MMS allows deploying multiple models on a single endpoint thanks to the multi-model endpoints functionality of Amazon SageMaker Hosting (https://docs.aws.amazon.com/sagemaker/latest/dg/multi-model-endpoints.html).

Useful links:
- https://github.com/awslabs/multi-model-server/
- https://github.com/aws/sagemaker-inference-toolkit

We start by defining some variables like the current execution role, the ECR repository that we are going to use for pushing the custom Docker container and a default Amazon S3 bucket to be used by Amazon SageMaker.

In [2]:
import boto3
import sagemaker
from sagemaker import get_execution_role

ecr_namespace = 'sagemaker-serving-containers/'
prefix = 'diffdl-container'

ecr_repository_name = ecr_namespace + prefix
role = get_execution_role()
account_id = role.split(':')[4]
region = boto3.Session().region_name
sagemaker_session = sagemaker.session.Session()
bucket = sagemaker_session.default_bucket()

print(account_id)
print(region)
print(role)
print(bucket)

785577973223
us-east-1
arn:aws:iam::785577973223:role/admin
sagemaker-us-east-1-785577973223


Let's take a look at the Dockerfile which defines the statements for building our serving container:

In [3]:
! pygmentize ../docker/Dockerfile

FROM tensorflow/tensorflow:latest

LABEL maintainer="Oleg Grytsynevych"

# Set a docker label to advertise multi-model support on the container
LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
# Set a docker label to enable container to use SAGEMAKER_BIND_TO_PORT environment variable if present
LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true

# Install python and other runtime dependencies
RUN apt-get update && \
    apt-get -y install \
        build-essential \
        libatlas-dev \
        git \
        wget \
        curl \
        openjdk-8-jdk-headless

# Python won’t try to write .pyc or .pyo files on the import of source modules
# Force stdin, stdout and stderr to be totally unbuffered. Good for logging
ENV PYTHONDONTWRITEBYTECODE=1 PYTHONUNBUFFERED=1 PYTHONIOENCODING=UTF-8 LANG=C.UTF-8 LC_ALL=C.UTF-8

# Install MMS, and SageMaker Inference Toolkit to set up MMS
RUN pip --no-cache-dir install multi-model-server \
                               s

At high-level the Dockerfile specifies the following operations for building this container:

- Set two Docker labels to advertise multi-model support and to enable the container using the SAGEMAKER_BIND_TO_PORT environment variable, if present
- Install libraries (including OpenJDK since MMS frontend is Java-based) and Python 3.6 through miniconda
- Set e few environment variables, including PYTHONUNBUFFERED which is used to avoid buffering Python standard output (useful for logging)
- Install XGBoost (it is the ML framework of choice for this example)
- Install Multi Model Server (MMS) and SageMaker Inference Toolkit
- Copy a .tar.gz package named <strong>multi_model_serving-1.0.0.tar.gz</strong> in the WORKDIR
- Install this package
- Copy the serve.py file in the WORKDIR and use it as the Docker ENTRYPOINT

Let's see the content of the <strong>serve.py</strong> file.

In [4]:
! pygmentize ../docker/code/serve.py

from __future__ import absolute_import

from subprocess import CalledProcessError
from retrying import retry
from multi_model_serving import handler_service
from sagemaker_inference import model_server

HANDLER_SERVICE = handler_service.__name__

def _retry_if_error(exception):
    return isinstance(exception, CalledProcessError)

@retry(stop_max_delay=1000 * 30,
       retry_on_exception=_retry_if_error)
def _start_model_server():
    # there's a race condition that causes the model server command to
    # sometimes fail with 'bad address'. more investigation needed
    # retry starting mms until it's ready
    model_server.start_model_server(handler_service=HANDLER_SERVICE)

if __name__ == '__main__':
    _start_model_server()


<h2>Handler Service</h2>

When looking at the Dockerfile above, you might be askiong yourself what the <strong>multi_model_serving-1.0.0.tar.gz</strong> package is.
When building a framework container for serving, sagemaker-inference-toolkit allows you to pass an handler service that will define the default inference handling logic, when users do not pass any custom inference script. The package above contains this code.

This is the content of the handler service:

In [5]:
! pygmentize ../package/src/multi_model_serving/handler_service.py

from sagemaker_inference.default_handler_service import DefaultHandlerService
from sagemaker_inference.transformer import Transformer
from multi_model_serving.default_inference_handler import DefaultDiffDLInferenceHandler

import os
import sys

ENABLE_MULTI_MODEL = os.getenv("SAGEMAKER_MULTI_MODEL", "false") == "true"

class HandlerService(DefaultHandlerService):
    def __init__(self):
        self._initialized = False
        
        transformer = Transformer(default_inference_handler=DefaultDiffDLInferenceHandler())
        super(HandlerService, self).__init__(transformer=transformer)

    def initialize(self, context):
        # This code is a workaround to fix a bug in the inference toolkit not setting the
        # user module path correctly when multi-model is enabled.
        # To be removed when the toolkit fix is available.
        if (not self._initialized) and ENABLE_MULTI_MODEL:
            code_dir = context.system_properties.get("model_dir") + '/code'
            sys.pa

And this is the logic defined in the default inference handler:

In [6]:
! pygmentize ../package/src/multi_model_serving/default_inference_handler.py

import pickle as pkl
from sagemaker_inference import content_types, decoder, default_inference_handler, encoder, errors
from multi_model_serving import encoder as xgb_encoders

class DefaultDiffDLInferenceHandler(default_inference_handler.DefaultInferenceHandler):

    def default_input_fn(self, input_data, content_type):
        """Take request data and de-serializes the data into an object for prediction.
        When an InvokeEndpoint operation is made against an Endpoint running SageMaker model server,
        the model server receives two pieces of information:
            - The request Content-Type, for example "application/json"
            - The request data, which is at most 5 MB (5 * 1024 * 1024 bytes) in size.
        The input_fn is responsible to take the request data and pre-process it before prediction.
        Args:
            input_data (obj): the request data.
            content_type (str): the request Content-Type.
        Returns:
            (obj): data ready for

<h2>Build and push the container</h2>
We are now ready to build this container and push it to Amazon ECR. This task is executed using a shell script stored in the ../script/ folder. Let's take a look at this script and then execute it.

In [15]:
! pygmentize ../scripts/build_and_push.sh

#!/bin/sh

ACCOUNT_ID=$1
REGION=$2
REPO_NAME=$3

cd ../package/ || exit 1
python setup.py sdist || exit 1
cp dist/multi_model_serving-1.0.0.tar.gz ../docker/code/ || exit 1

docker build -f ../docker/Dockerfile -t $REPO_NAME ../docker

docker tag $REPO_NAME $ACCOUNT_ID.dkr.ecr.$REGION.amazonaws.com/$REPO_NAME:latest

$(aws ecr get-login --no-include-email --registry-ids $ACCOUNT_ID)

aws ecr describe-repositories --repository-names $REPO_NAME || aws ecr create-repository --repository-name $REPO_NAME

docker push $ACCOUNT_ID.dkr.ecr.$REGION.amazonaws.com/$REPO_NAME:latest


<h3>--------------------------------------------------------------------------------------------------------------------</h3>

The script builds the Docker container, then creates the repository if it does not exist, and finally pushes the container to the ECR repository. The build task requires a few minutes to be executed the first time, then Docker caches build outputs to be reused for the subsequent build operations.

In [8]:
!sudo yum -y install docker
!/opt/conda/bin/pip install docker-compose
!/opt/conda/bin/pip install 'sagemaker[local]'

Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
amzn2-core                                               | 3.7 kB     00:00     
242 packages excluded due to repository priority protections
Package docker-20.10.4-1.amzn2.x86_64 already installed and latest version
Nothing to do


In [ ]:
! ../scripts/build_and_push.sh $account_id $region $ecr_repository_name

running sdist
running egg_info
writing src/multi_model_serving.egg-info/PKG-INFO
writing top-level names to src/multi_model_serving.egg-info/top_level.txt
writing dependency_links to src/multi_model_serving.egg-info/dependency_links.txt
reading manifest file 'src/multi_model_serving.egg-info/SOURCES.txt'
writing manifest file 'src/multi_model_serving.egg-info/SOURCES.txt'

running check

creating multi_model_serving-1.0.0
creating multi_model_serving-1.0.0/src
creating multi_model_serving-1.0.0/src/multi_model_serving
creating multi_model_serving-1.0.0/src/multi_model_serving.egg-info
copying files to multi_model_serving-1.0.0...
copying setup.py -> multi_model_serving-1.0.0
copying src/multi_model_serving/__init__.py -> multi_model_serving-1.0.0/src/multi_model_serving
copying src/multi_model_serving/default_inference_handler.py -> multi_model_serving-1.0.0/src/multi_model_serving
copying src/multi_model_serving/encoder.py -> multi_model_serving-1.0.0/src/multi_model_serving
copying s

Get:27 http://archive.ubuntu.com/ubuntu bionic/main amd64 fonts-dejavu-core all 2.37-1 [1041 kB]
Get:28 http://archive.ubuntu.com/ubuntu bionic/main amd64 fontconfig-config all 2.12.6-0ubuntu2 [55.8 kB]
Get:29 http://archive.ubuntu.com/ubuntu bionic/main amd64 libfontconfig1 amd64 2.12.6-0ubuntu2 [137 kB]
Get:30 http://archive.ubuntu.com/ubuntu bionic/main amd64 libnspr4 amd64 2:4.18-1ubuntu1 [112 kB]
Get:31 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libnss3 amd64 2:3.35-2ubuntu2.12 [1220 kB]
Get:32 http://archive.ubuntu.com/ubuntu bionic/main amd64 libpcsclite1 amd64 1.8.23-1 [21.3 kB]
Get:33 http://archive.ubuntu.com/ubuntu bionic/main amd64 libxi6 amd64 2:1.7.9-1 [29.2 kB]
Get:34 http://archive.ubuntu.com/ubuntu bionic/main amd64 libxrender1 amd64 1:0.9.10-1 [18.7 kB]
Get:35 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 x11-common all 1:7.7+19ubuntu7.1 [22.5 kB]
Get:36 http://archive.ubuntu.com/ubuntu bionic/main amd64 libxtst6 amd64 2:1.2.3-1 [12.8 kB]


Selecting previously unselected package x11-common.
Preparing to unpack .../27-x11-common_1%3a7.7+19ubuntu7.1_all.deb ...
dpkg-query: no packages found matching nux-tools
Unpacking x11-common (1:7.7+19ubuntu7.1) ...
Selecting previously unselected package libxtst6:amd64.
Preparing to unpack .../28-libxtst6_2%3a1.2.3-1_amd64.deb ...
Unpacking libxtst6:amd64 (2:1.2.3-1) ...
Selecting previously unselected package openjdk-8-jre-headless:amd64.
Preparing to unpack .../29-openjdk-8-jre-headless_8u292-b10-0ubuntu1~18.04_amd64.deb ...
Unpacking openjdk-8-jre-headless:amd64 (8u292-b10-0ubuntu1~18.04) ...
Selecting previously unselected package ca-certificates-java.
Preparing to unpack .../30-ca-certificates-java_20180516ubuntu1~18.04.1_all.deb ...
Unpacking ca-certificates-java (20180516ubuntu1~18.04.1) ...
Selecting previously unselected package libcurl3-gnutls:amd64.
Preparing to unpack .../31-libcurl3-gnutls_7.58.0-2ubuntu3.13_amd64.deb ...
Unpacking libcurl3-gnutls:amd64 (7.58.0-2ubuntu3.1

Adding debian:Comodo_AAA_Services_root.pem
Adding debian:Cybertrust_Global_Root.pem
Adding debian:D-TRUST_Root_Class_3_CA_2_2009.pem
Adding debian:D-TRUST_Root_Class_3_CA_2_EV_2009.pem
Adding debian:DST_Root_CA_X3.pem
Adding debian:DigiCert_Assured_ID_Root_CA.pem
Adding debian:DigiCert_Assured_ID_Root_G2.pem
Adding debian:DigiCert_Assured_ID_Root_G3.pem
Adding debian:DigiCert_Global_Root_CA.pem
Adding debian:DigiCert_Global_Root_G2.pem
Adding debian:DigiCert_Global_Root_G3.pem
Adding debian:DigiCert_High_Assurance_EV_Root_CA.pem
Adding debian:DigiCert_Trusted_Root_G4.pem
Adding debian:E-Tugra_Certification_Authority.pem
Adding debian:EC-ACC.pem
Adding debian:Entrust.net_Premium_2048_Secure_Server_CA.pem
Adding debian:Entrust_Root_Certification_Authority.pem
Adding debian:Entrust_Root_Certification_Authority_-_EC1.pem
Adding debian:Entrust_Root_Certification_Authority_-_G2.pem
Adding debian:Entrust_Root_Certification_Authority_-_G4.pem
Adding debian:GDCA_TrustAUTH_R5_ROOT.pem
Adding deb

Processing triggers for ca-certificates (20210119~18.04.1) ...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...

done.
done.
Processing triggers for mime-support (3.60ubuntu1) ...
Removing intermediate container 3fc843192116
 ---> 95d4f2cec3cb
Step 6/15 : ENV PYTHONDONTWRITEBYTECODE=1 PYTHONUNBUFFERED=1 PYTHONIOENCODING=UTF-8 LANG=C.UTF-8 LC_ALL=C.UTF-8
 ---> Running in 83bd91206e60
Removing intermediate container 83bd91206e60
 ---> b1531e12072b
Step 7/15 : RUN mkdir -p /opt/ml/model
 ---> Running in 3a8045c0c0f1
Removing intermediate container 3a8045c0c0f1
 ---> eb4bd133c323
Step 8/15 : RUN pip --no-cache-dir install -U pip
 ---> Running in c14bd3fd5ef2
  Attempting uninstall: pip
    Found existing installation: pip 20.2.4
    Uninstalling pip-20.2.4:
      Successfully uninstalled pip-20.2.4
Removing intermediate container c14bd3fd5ef2
 ---> 404b20b741ca
Step 9/15 : RUN pip --no-cache-dir install multi-model-serve

The push refers to repository [785577973223.dkr.ecr.us-east-1.amazonaws.com/sagemaker-serving-containers/diffdl-container]

897bce40: Preparing 
dae4b8dd: Preparing 
87f9a136: Preparing 
c1070646: Preparing 
3f9722a9: Preparing 
633053c1: Preparing 
fa7296bc: Preparing 
7734672b: Preparing 
0c0a3deb: Preparing 
1cfe71d6: Preparing 
d69c1c3b: Preparing 
d7a6e6b8: Preparing 
828d7364: Preparing 
b79b7c2f: Preparing 
763a045a: Preparing 
c6d2db45: Preparing 
bacb0351: Preparing 
1070646: Pushing  30.49MB/84.46MB7A

<h2>Deploy with Amazon SageMaker</h2>


<h3>Get the container URI</h3>
Once we have correctly pushed our container to Amazon ECR, we are ready to deploy with Amazon SageMaker, which requires the ECR path to the Docker container used for serving as parameter for deployment.

In [14]:
container_image_uri = '{0}.dkr.ecr.{1}.amazonaws.com/{2}:latest'.format(account_id, region, ecr_repository_name)
print(container_image_uri)

785577973223.dkr.ecr.us-east-1.amazonaws.com/sagemaker-serving-containers/diffdl-container:latest


<h3>Prepare two models</h3>

We are going to deploy two different XGBoost models to our model server. We will need the serialized models and the inference scripts that we want to use.
We will store them in the current notebook folder, under <strong>model_and_code_1/</strong> and <strong>model_and_code_2/</strong>.

The purpose of using different models is to show that you can also deploy models that require diverse features and pre/post processing code.

First model is a regression model trained on the [Abalone data](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/regression.html) originally from the [UCI data repository](https://archive.ics.uci.edu/ml/datasets/abalone).
For further information, please refer to this [example](https://github.com/awslabs/amazon-sagemaker-examples/blob/master/introduction_to_amazon_algorithms/xgboost_abalone/xgboost_abalone.ipynb).

Second model is a binary classification model built by following this workshop: https://github.com/aws-samples/amazon-sagemaker-build-train-deploy

In [16]:
! rm -rf ./model_and_code_1/.ipynb_checkpoints
! rm -rf ./model_and_code_1/code/.ipynb_checkpoints
! rm -rf ./model_and_code_2/.ipynb_checkpoints
! rm -rf ./model_and_code_2/code/.ipynb_checkpoints

! tar -C ./model_and_code_1/ -cvzf model1.tar.gz ./
! tar -C ./model_and_code_2/ -cvzf model2.tar.gz ./

./
./code/
./code/predictor.py
./xgboost-model
./
./code/
./code/predictor.py
./model.bin


Let's see the custom inference script for the first model:

In [10]:
! pygmentize model_and_code_1/code/predictor.py

import os
import pickle as pkl

def model_fn(model_dir):
    model_file = model_dir + '/xgboost-model'
    model = pkl.load(open(model_file, 'rb'))
    return model


And this is the one for the second model:

In [11]:
! pygmentize model_and_code_2/code/predictor.py

import os
import pickle as pkl

from multi_model_serving import encoder as xgb_encoders

def input_fn(input_data, content_type):    
    return xgb_encoders.decode(input_data, content_type)

def model_fn(model_dir):
    model_file = model_dir + '/model.bin'
    model = pkl.load(open(model_file, 'rb'))
    return model


<h3>Deploy a single model</h3>

In [17]:
s3_model_path = 's3://{0}/{1}/model/model1.tar.gz'.format(bucket, prefix)
!aws s3 cp model1.tar.gz {s3_model_path}

upload: ./model1.tar.gz to s3://sagemaker-us-east-1-785577973223/diffdl-container/model/model1.tar.gz


In [19]:
from time import gmtime, strftime
from sagemaker.model import Model

model_name = 'multi-model-server-model-' + strftime("%Y-%m-%d-%H-%M-%S", gmtime())

model = Model(model_data = s3_model_path,
              image_uri = container_image_uri,
              env = {
                  'SAGEMAKER_PROGRAM': 'predictor'
              },
              role=role,
              name = model_name,
              predictor_cls = sagemaker.predictor.Predictor,
              #sagemaker_session=sagemaker_session #comment this line for local mode.
             )

<strong>Note:</strong> the environment variable SAGEMAKER_PREDICTOR is used to specify the name of the custom inference script.

Test the image locally `docker run -it --rm 785577973223.dkr.ecr.us-east-1.amazonaws.com/sagemaker-serving-containers/diffdl-container:latest bash`.
Inference API bind to: http://0.0.0.0:8080

In [ ]:
(echo -n '{ INSERT JSON HERE }') | curl -H "Content-Type: application/json" -d @- $ENDPOINT_URL

In [23]:
endpoint_name = 'multi-model-server-single-ep-' + strftime("%Y-%m-%d-%H-%M-%S", gmtime())
print(endpoint_name)
pred = model.deploy(initial_instance_count=1,
                    instance_type='local',
                    endpoint_name=endpoint_name)

Using the short-lived AWS credentials found in session. They might expire while running.
Failed to import yaml. Local mode features will be impaired or broken. Please run "pip install 'sagemaker[local]'" to install all required dependencies.


multi-model-server-single-ep-2021-07-15-22-06-47


ModuleNotFoundError: No module named 'yaml'

In [ ]:
from sagemaker.predictor import Predictor

pred.serializer = sagemaker.serializers.CSVSerializer()
item = '77,33,143.0,101,212.2,102,104.9,120,15.3,4,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,1'
pred.predict(item)

In [ ]:
pred.delete_endpoint()
pred.delete_model()

<h3>Deploy multiple models</h3>

In [ ]:
model_data_prefix = 's3://{0}/{1}/modeldata'.format(bucket, prefix)

s3_model_1_path = model_data_prefix + '/model1.tar.gz'
!aws s3 cp model1.tar.gz {s3_model_1_path}
s3_model_2_path = model_data_prefix + '/model2.tar.gz'
!aws s3 cp model2.tar.gz {s3_model_2_path}

In [ ]:
from time import gmtime, strftime
from sagemaker.multidatamodel import MultiDataModel
from sagemaker.model import Model

model_name = 'multi-model-server-multidatamodel-' + strftime("%Y-%m-%d-%H-%M-%S", gmtime())

model = Model(name = model_name,
              model_data = '',
              image_uri = container_image_uri,
              role=role,
              env = {
                  'SAGEMAKER_PROGRAM': 'predictor'
              },
              predictor_cls = sagemaker.predictor.Predictor,
              sagemaker_session=sagemaker_session)

multi_model = MultiDataModel(name = model_name,
                             model_data_prefix = model_data_prefix,
                             model = model,
                             sagemaker_session=sagemaker_session)

<strong>Note:</strong> the environment variable SAGEMAKER_PREDICTOR is used to specify the name of the custom inference script.

In [ ]:
multi_endpoint_name = 'multi-model-server-ep-' + strftime("%Y-%m-%d-%H-%M-%S", gmtime())
print(multi_endpoint_name)

pred = multi_model.deploy(initial_instance_count=1,
                          instance_type='ml.m5.xlarge',
                          endpoint_name=multi_endpoint_name)

<h3>Executing inferences</h3>
Once the multi-model endpoint is ready, we can invoke either model1 or model2 by changing the target_model variable in the predict() function call.

In [ ]:
from sagemaker.predictor import Predictor
pred = Predictor(multi_endpoint_name)
pred.serializer = sagemaker.serializers.CSVSerializer()

In [ ]:
item = '77,33,143.0,101,212.2,102,104.9,120,15.3,4,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,1'
model_archive = '/model1.tar.gz'
pred.predict(item, target_model=model_archive)

In [ ]:
item = '0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,73.0,79.0,32.0,27.0,45.0,48.0,13.0,62.0'
model_archive = '/model2.tar.gz'
pred.predict(item, target_model=model_archive)

In [ ]:
pred.delete_endpoint()
pred.delete_model()